src/Service/CaptureServiceImpl.cpp

```diff
@@ -216,6 +216,11 @@ grpc::Status CaptureServiceImpl::Capture(
   LOG("Read CaptureRequest from Capture's gRPC stream: starting capture");
 
   tracing_handler.Start(request.capture_options());
+
+  // Tracing handler is responsible for generating CaptureStarted event which should always
+  // go first, we wait until it is done before continuing.
+  tracing_handler.WaitForCaptureStartedEvent();
```

Comment:
I think there's a discussion to be had here about `CaptureStarted` being generated by `LinuxTracing`. This effectively makes `LinuxTracing` the "master" producer. Secondarily, it also adds some code to `TracerThread::Startup` that doesn't necessarily belong there (although there's already so much stuff in `LinuxTracing` anyway).
This doesn't seem very clean. What's the reason for this? Can't `CaptureStarted` be generate in a dedicated place? The only thing that really seems to belong in `LinuxTracing` is the initial thread names, but that can be done differently, e.g., the same way it was done before or with a dedicated `InitialThreadNames` message.
Have you considered alternatives?

Reply:
I do not think initial thread names should be sent separately, we should have a snapshot of a process state at the start of the capture. It makes sense to include this to CaptureStartedEvent.

I was considering moving compilation of CaptureSartedEvent to ProducerEventProcessor but this is much bigger change. I was thinking that the initial memory snapshot should also be included to CaptureStarted event. And decided to postpone this to a separate change as well, this is because we might want to consider having initial messages for all other producers, this in will have affect on delaying all other messages sent to the client until all Start messages gathered and sent. And I am not sure we should do that since this could have a negative impact on memory consumption in the service and might create latency problems.

Do you have anything in particular you want to be addressed?

Reply:
> I do not think initial thread names should be sent separately, we should have a snapshot of a process state at the start of the capture.

This was a minor detail. 

But I didn't really get an opinion on the main point, which is architectural and was the following: "This effectively makes LinuxTracing the "master" producer." In other words, this makes `LinuxTracing` indispensable and it makes all other producers wait for part of its initialization (dropping all their previous events).

This is a very strong requirement with potential future implications (e.g., a capture that collects data only from the Vulkan layer is not a valid one). So I want to understand if and why it's not possible to do it differently (for example, dedicated code separated from `LinuxTracing`, dedicated producer, etc).

Reply:
we now have snapshot events.

Reply:
and now we dont.

Reply:
and now we do again.